### ISSUE TITLE
SCAN, SSCAN, HSCAN, and ZSCAN are not stable across failovers

### ISSUE THREAD

**Issue by madolson** [madolson/placeholderkv#4]
Scan is supposed to provide the following guarantees (as per https://redis.io/docs/manual/keyspace/):

A full iteration always retrieves all the elements that were present in the collection from the start to the end of a full iteration. This means that if a given element is inside the collection when an iteration is started, and is still there when an iteration terminates, then at some point [SCAN](https://redis.io/commands/scan) returned it to the user.

A full iteration never returns any element that was NOT present in the collection from the start to the end of a full iteration. So if an element was removed before the start of an iteration, and is never added back to the collection for all the time an iteration lasts, [SCAN](https://redis.io/commands/scan) ensures that this element will never be returned.

While playing around with some PoC for cluster wide scan, I realized that the first guarantee can only hold as long as you were connected to the same node during the entire duration of the scanning. If failover occurs, the replica may have a different seed value for the siphash function, so the cursor previously used on the primary would not be in the same place. This is likely less of an issue for SCAN since you normally indicate the node you're talking to, but many CME client transparently handle re-directs for SSCAN/HSCAN/ZSCAN during failovers.

I'm not sure we need to strictly do anything about SCAN, I'm not sure how often SCAN is resumed after a failover. The other commands might though.

I think this is worth addressing, but could be done in three ways:

Simply update the documentation to add the caveat. This doesn't feel right to me, because most users will not really be aware of this.
Add a configuration so that seeds can be set externally. This would allow operators to configure this consistency, but has limited benefit for those that don't know.
Allow replicas to sync their data seed from their primaries. This makes their cursors consistent. We can also persistent this into RDB so that it is still accurate, I'm most in-favor of this.

ref: https://github.com/redis/redis/issues/12440

We never came up with a cohesive answer here

**Comment by madolson** [comment]
There is a second issue that I want to make sure we don't drop, which is the cursor format is changing for SCAN for Redis 8 with the new per-slot dictionaries. The format is changing from <64 bits for DB cursor> to 00<14 bits for slot><48 bits for DB cursor>. To be specific, we aren't actually showing the bits, just the integer representation of those bits. We should be able to detect versioning issues, i.e. using the cursor from 7.0 node on an 8.0 node, so I would like to propose we update it to one of the following two options:

<2 version bits><14 bits for slot><48 bits for DB cursor>, we will bump from version 00 -> 01. This is likely the most backwards compatible. It also allows us in the future to do a third version if we want to re-organize the cursor bits. Going to more than 2 bits introduces the risk there are users are storing the cursor as a long long, and it will break.
-<14 bits for slot><48 bits for DB cursor>. The version will initially be 1, and we'll use the - to detect the new format. This gives us the most freedom to change the version more in the future.
@yossigo, I wasn't able to find our decision from the previous meeting. I know you were concerned with the format, which we agreed we should look into fixing, but I don't recall if you also wanted to try to make the SCAN command stable across failover. (I still want to try to make HSCAN, ZSCAN, and SSCAN stable) So that it might return duplicate items, but won't omit items because of the cursor shift.

**Comment by zuiderkwast** [comment]
This is a great topic to finish. So it's actually three things mentioned here:

1. Scan across failovers. Synching the hashtable seed. We can do that in the RDB or in another way when initiating replication or before. The seed may be sensitive security information though, but if anyone has permission to replicate, I think it's fine to let them fetch the seed too.
2. Cluster-wide scan. Per slot, either explicitly or implicitly. You were against exposing the slot to the user, while others liked it.
3. Cursor versioning. I don't see how you can use these bits though. Are these two LSB unused in an old cursor?

**Comment by xingbowang** [comment]
I talked with Madelyn offline. It seems the root cause of this is issue is different hash-seed was used on different nodes. I propose to use a boot-time config parameter "hash-seed-override" to override the randomly generated seed in the engine. This allows us to keep the default behavior unchanged. Meantime, if someone would like to fix this issue, they could add this config in the bootstrap parameter in their engine bootstrap configuration. Since the bootstrap configuration is used in every deployment of Valkey, it should not be a big effort to add an extra config parameter at boot time. The format could be something like.
```
config set hash-seed-override $SEED_IN_UINT64_T
```
Sample
```
config set hash-seed-override 1234567
```

**Comment by sarthakaggarwal97** [comment]
I started thinking about this issue, most importantly, two of the three approaches that @madolson mentioned in the description. 

**Approaches:** 

1. Add a configuration so that seeds can be set externally. This would allow operators to configure this consistency, but has limited benefit for those that don't know.
2. Allow replicas to sync their data seed from their primaries. This makes their cursors consistent. We can also persistent this into RDB so that it is still accurate.


**For Approach 1**, we can introduce a `hash-seed-override` configuration so all nodes (within the shard / cluster) can be started with the same seed. If every primary and replica shares the seed, iteration order becomes stable across failover. This is simple to implement and opt-in, with no compatibility break. 
The concern could be around having a similar seed per shard, or per cluster, but as I understand `siphash` is a secure PRF, and avoid any DOS type attacks.


**For Approach 2** we can propagate the primary’s `hash-seed` to replicas so a failover can preserve iteration order without operator intervention. The seed can be written and read through RDB AUX metadata and applied before any keys are loaded. The full sync should automatically deliver the seed to new replicas. On restart with persisted data, the primary reloads the same seed from RDB/AOF, keeping replicas consistent. Partial syncs are only safe as long as both sides retain the same seed. 

I believe there would be multiple considerations here, I thought through some of the cases like scale in/out, upgrades, failovers, diskless replication etc, so sharing some thoughts here:
1. It takes away the operator pain for configuration, but it is a bit complicated to implement and would have significant blast radius.
2. Adding or Removing a Replica: Most importantly, in scenarios of upgrades, when the cluster is of mixed version, if the replica doesn’t get that hash-seed from old primary, it will generate it’s own seed and diverge. When finally the failover would occur to new primary, all the nodes' seeds would have diverged, and we will have to do full sync again. 
3. Failovers: In the rare scenarios where replica is not fully synced (e.g., only partial data, or promotion happens mid-sync) and kept a different seed after becoming primary. This would leave us with similar problems again. When this replica is promoted to primary, it will try to share it’s own seed via replication stream, which would require a full sync across the shard.
4. @zuiderkwast raised a concern around the security aspect of sharing the seed through replication stream, but I am not totally sure on this. 

I propose to move with Approach 1. I do agree that Approach 1 solves the problem only for those who know, but reduces the variables of where things can go wrong while giving a way out for those who are facing the issue. With Approach 2, I am not sure if there's a way to fully avoid the ambiguities of real world scenarios, and I am worried if one thing can break other functionalities.

**Comment by sarthakaggarwal97** [comment]
> Cursor versioning. I don't see how you can use these bits though. Are these two LSB unused in an old cursor?

I might be wrong here, and it's not documented as well. But if I understand correctly, we have the following structure currently.

```
[ 48-bit hashtable cursor ][ 14-bit for slots]
```

So we might have last two LSB's unused here.

**Comment by sarthakaggarwal97** [comment]
@zuiderkwast @madolson do you guys have any suggestions on how should we approach this?

**Comment by madolson** [comment]
I guess I'm aligned to just do 1.

**Comment by zuiderkwast** [comment]
1 is definitely easier, but 2 provides better guarantees. Can the client authors (say Glide) trust that this seed is configured correctly on all nodes? If it isn't, we don't get the guarantees and then maybe clients won't be able to use this?

I guess I'm OK with 1, but if we manage to solve 2, it becomes fully automatic, more convenient and we can get the guarantees. Let's go over the challanges about approach 2 first?

> 2\. Adding or Removing a Replica: Most importantly, in scenarios of upgrades, when the cluster is of mixed version, if the replica doesn’t get that hash-seed from old primary, it will generate it’s own seed and diverge. When finally the failover would occur to new primary, all the nodes' seeds would have diverged, and we will have to do full sync again.

If we can propagate the seed, can we trigger rehashing using the new seed without doing a full sync? A replica can fetch it from the primary using replconf before psync. We need to store the seed (or hash function) per table but I guess it's doable.

> 3\. Failovers: In the rare scenarios where replica is not fully synced (e.g., only partial data, or promotion happens mid-sync) and kept a different seed after becoming primary. This would leave us with similar problems again. When this replica is promoted to primary, it will try to share it’s own seed via replication stream, which would require a full sync across the shard.

Same solution as above?

> 4\. [@zuiderkwast](https://github.com/zuiderkwast) raised a concern around the security aspect of sharing the seed through replication stream, but I am not totally sure on this.

I'm not worried about it. We should just make sure unauthenticated clients can't access the seed.

**Comment by sarthakaggarwal97** [comment]
> 1 is definitely easier, but 2 provides better guarantees. Can the client authors (say Glide) trust that this seed is configured correctly on all nodes? If it isn't, we don't get the guarantees and then maybe clients won't be able to use this?

I think the clients can trust if the seed is provided by configuration, but otherwise we don't really expose this seed value to verify iirc.

> can we trigger rehashing using the new seed without doing a full sync?

I am worried that reshashing would be an expensive operation. We will be pretty much rehashing whenever we do slot migration. I was talking to @madolson on this offline, and we were kinda thinking that it's a lot of work and handling to handle a single edge case. Wdyt?

**Comment by zuiderkwast** [comment]
> I am worried that reshashing would be an expensive operation.

It is, but it's amortized over time and it's triggered only in rare cases: failover in mixed cluster, i.e. during rolling upgrade, and failover before full sync is complete, in which case you'd have worse problems (data loss) so I don't really buy this concern. Rehasing happens from time to time in any node when the data grows.

> We will be pretty much rehashing whenever we do slot migration.

I don't understand this. During slot migration, each key is inserted in the target using the target's seed. 🤔 Do you mean we want scan to be stable cross slot migrations too? It makes sense, though it wasn't mentioned before in this issue. :)

I can't see we need rehashing for this. I imagine we'd need a seed per slot, to propagate it in slot migration and to store them all in RDB (aux fields).

Am I missing something?

**Comment by madolson** [comment]
> I can't see we need rehashing for this. I imagine we'd need a seed per slot, to propagate it in slot migration and to store them all in RDB (aux fields).

The current implementation doesn't support one seed per hashtable, it's a global for all hashtables. If we wanted to make it metadata, it's an extra 8 bytes-ish of data for each hashtable. When we do a slot migration, we either need to re-hash "All" slots in the cluster to conform to the new seed (possible) or add the extra metadata per hashtable. Keep in mind this also applies to underlying items as well, so HSCAN upholds the requirements.

I would like us to build something that is seamless. It just seems like a lot of complexity.

**Comment by zuiderkwast** [comment]
> If we wanted to make it metadata, it's an extra 8 bytes-ish of data for each hashtable.

Yeah, the hashtable struct would grow from 64 bytes to 80 (two seeds, one per table). We use listpacks for small objects, but still.

> When we do a slot migration, we either need to re-hash "All" slots in the cluster to conform to the new seed (possible)

I'm not sure that works. Whenever we rehash a table to a different seed, we break any ongoing scan on that table. It seems OK in a mixed cluster (during upgrade) but not on every slot migration. It seems to me that the only way is seed per table.

> I would like us to build something that is seamless. It just seems like a lot of complexity.

Yeah, it does get very complex and costs extra memory. I'm convinced. Let's do the config.

**Comment by madolson** [comment]
> Yeah, it does get very complex and costs extra memory. I'm convinced. Let's do the config.

Worth mentioning, if we ever come up with a nice solution later, we can always deprecate the config and use the new mechanism. So this is a reversible decisions.

**Comment by zuiderkwast** [comment]
> So this is a reversible decisions.

Yeah, but a deprecated config leaves a small footprint forever, a very small one though. :)
